### PR TITLE
Recognize self-closing tags

### PR DESCRIPTION
**Pull request by NLBuescher**
So I've been taking a look at issue #203 and I've fixed the issue of self-closing tags getting matched improperly.

I've modified the regular expressions throughout the `tag-finder` script to properly match tags based on the additional criterion of self-closing. I've also updated the `startRange` and `endRange` calculations accordingly. See below for an explanation of the issue and its side-effect.

-----

the bracketmatcher thinks that self-closing tags are actually opening tags and matches them with the next closing tag it can find as illustrated below:
test_space

Interestingly enough, it doesn't do this if you write the tag without a space between the tag name and the closing bit:
test_nospace

Problematically, it doesn't actually recognize the tag as a self-closing tag, but rather thinks that the tag name is "TEST/", and subsequently can't find a matching closing tag. This leads to the very interesting phenomenon shown below:
test_nospace_endtag

This problem is caused by the `tag-finder.coffee` script and its method of identifying tags:

```coffeescript
/(<(\/?))([^\s>]+)([\s>]|$)/
```

The above regex matches tags by finding the prefix `<`, with the optional modifier of `/` to make it a closing tag, then isolates the tag name by matching anything that isn't a closing angle bracket or whitespace`[^\s>]+`, then matches either a whitespace, a closing angle bracket, OR the EOL `[\s>]|$`. The problem with this is that it doesn't actually check all the way to the end of the tag to see if it closes itself, leading to the assumption that any tag that any whitespace after the tagname means the tag is an opening tag.

### CODE REVIEW

File: lib/tag-finder.coffee [hunk 1]
```diff
@@ -8,14 +8,14 @@ SelfClosingTags = require './self-closing-tags'
 module.exports =
 class TagFinder
   constructor: (@editor) ->
-    @tagPattern = /(<(\/?))([^\s>]+)([\s>]|$)/
-    @wordRegex = /[^>\r\n]*/
+    @tagPattern = /(<(\/?))([^\s>]+)([^/>]*)(\/?)(>)/
+    @wordRegex = /[^>\r\n]*>/
```
Comment (50Wliu) [inline lib/tag-finder.coffee]: Why was this changed?

**Comment by 50Wliu** [comment]
Can you please add a spec for this and also fix the failing one?

File: lib/tag-finder.coffee (hunk 1 above)
Comment (NLBuescher) [inline lib/tag-finder.coffee]: So the issue that I'm trying to fix is that "custom" self-closing tags, like a self-closing `span` element would not get identified properly by the `tag-finder.coffee` script and get treated the same as opening tags. I realized this was because the `@tagPattern` RegEx that found the tags only matched tags up to the `space` right before the attributes (class, id, style, etc.).

In other words the `tag-finder` was not going all the way past the attributes to the end of the tag, the closing angle bracket `>`, when identifying tags.

However, going all the way past the attributes to see whether the tag ends in `/>` or just a `>` was essential to determining whether or not the tag closes itself, and based on that, whether or not it should be matched with other closing tags.

**Comment by NLBuescher** [comment]
@50Wliu Sorry if I'm making mistakes, this is my first time submitting my pull request like this 😅. I'm gonna look into the failing checks and add the test case I'm trying to fix to the specs.

**Comment by NLBuescher** [comment]
I've fixed the failing checks and added two specs to test for the new use case.

File: spec/bracket-matcher-spec.coffee [hunk 2]
```diff
@@ -242,6 +242,26 @@ describe "bracket matching", ->
...
+          buffer.setText """
+            <test>
+              <test />
+            </test>
+          """
+
+          editor.setCursorBufferPosition([2, Infinity])
+          expectHighlights([2, 2], [0, 1])
+
+      describe "when on a self-closing tag", ->
```
Comment (50Wliu) [inline spec/bracket-matcher-spec.coffee]: I would like two extra tests:
1. Test to make sure `<test/>` is identified as a self-closing tag
2. Test to make sure `<test______/>` (where `_` is a space) is identified as a self-closing tag

File: lib/tag-finder.coffee [hunk 3]
```diff
@@ -8,8 +8,8 @@ SelfClosingTags = require './self-closing-tags'
 module.exports =
 class TagFinder
   constructor: (@editor) ->
-    @tagPattern = /(<(\/?))([^\s>]+)([^/>]*)(\/?)(>)/
-    @wordRegex = /[^>\r\n]*>/
+    @tagPattern = /(<(\/?))([^/\s>]+)([^/>]*)(\/?)(>)/
+    @wordRegex = /(?![\n\r])[^>]*>/
```
Comment (50Wliu) [inline lib/tag-finder.coffee]: Two things:
1. Can the `(?![\n\r])` be changed to simply be `(?!$)`?
2. Either way, please add the lookahead at the end of the match.

**Comment by 50Wliu** [comment]
> Sorry if I'm making mistakes, this is my first time submitting my pull request like this 😅

No problem.  You had a great PR writeup, which is a lot more than most PRs.

File: spec/bracket-matcher-spec.coffee (hunk 2 above)
Comment (NLBuescher) [inline spec/bracket-matcher-spec.coffee]: I've added the requested tests.

File: lib/tag-finder.coffee (hunk 3 above)
Comment (NLBuescher) [inline lib/tag-finder.coffee]: 1. The `(?![\n\r])` CAN be changed to `(?!$)`, thank you for the suggestion
2. So what I'm trying to achieve with the negated lookahead at the start of the match was to only match the string if it was not ***preceded*** by a newline or carriage return. In other words, what looks like a negated lookahead is actually, functionally, a negated lookbehind, and is placed exactly where it needs to be to make the whole thing work. (the tests start failing if I mess with it)

**Comment by NLBuescher** [comment]
Suggested changes were made where possible.

File: lib/tag-finder.coffee (hunk 3 above)
Comment (50Wliu) [inline lib/tag-finder.coffee]: Ah, ok.

**Comment by baldurh** [comment]
Can’t wait for this one!

File: lib/tag-finder.coffee [hunk 4]
```diff
@@ -8,14 +8,14 @@ SelfClosingTags = require './self-closing-tags'
 module.exports =
 class TagFinder
   constructor: (@editor) ->
-    @tagPattern = /(<(\/?))([^\s>]+)([\s>]|$)/
-    @wordRegex = /[^>\r\n]*/
+    @tagPattern = /(<(\/?))([^\s>]+)([^/>]*)(\/?)(>)/
```
Comment (MoritzKn) [inline lib/tag-finder.coffee]: I don't know how to fix this, but `/(<(\/?))([^\s>]+)([^/>]*)(\/?)(>)/` won't match `<foo bar="/">`. This is because `[^/>]` dose not mean "not the sequence `/>`" but match all characters but not `/` or `>`. Same goes for the regex in `patternForTagName`.

File: lib/tag-finder.coffee [hunk 5]
```diff
@@ -8,14 +8,14 @@ SelfClosingTags = require './self-closing-tags'
 module.exports =
 class TagFinder
   constructor: (@editor) ->
-    @tagPattern = /(<(\/?))([^\s>]+)([\s>]|$)/
-    @wordRegex = /[^>\r\n]*/
+    @tagPattern = /(<(\/?))([^/\s>]+)([^/>]*)(\/?)(>)/
```
Comment (MoritzKn) [inline lib/tag-finder.coffee]: The regex `(<(\/?))([^/\s>]+)([^/>]*)(\/?)(>)` won't match `<foo bar="/">`. This is because `[^/>]` doesn't mean "not the sequence `/>`", but match all characters with the exception of `/` and `>`. Same goes for the regex in `patternForTagName`. Maybe `(<(\/?))([^/\s>]+)((?:[^/>]|[^>](?!>))*)(\/?)(>)` could be used instead. In this regex the pattern `[^/>]*` is replaces by `((?:[^/>]|[^>](?!>))*`, which will again match all characters with the exception of `/` and `>`, but also the character `/` if it isn't followed by the character `>`.

Comment (NLBuescher) [inline lib/tag-finder.coffee]: That's a good catch. I'll look into that right away.

Comment (NLBuescher) [inline lib/tag-finder.coffee]: for reference, the test html I'm using:

    <TEST>
        <foo bar="/">
        </foo>
        <TEST/>
    </TEST>

so your suggestion does help with recognizing a tag like `<foo bar="/">`, but introduces a weird problem where, if I implement the fix on both `@tagPattern` and in `patternForTagName`, then putting the cursor on the last closing tag `</TEST>` doesn't highlight that pair (it highlights fine with the opening tag).

If I replace only the `@tagPattern` sequence, the program doesn't match the closing `</foo>` in the same manner as the closing `</TEST>`.

Looking into this problem now.

Comment (MoritzKn) [inline lib/tag-finder.coffee]: In my comment I originally said:
> In this regex the pattern `[^/>]*` is replaces by `((?:[^/>]|[^>](?!>))*`

this was wrong, I meant `(?:[^/>]|[^>](?!>))*` (without the first bracket). I just fixed that in the comment. If you copied that, this may have caused the problem...
If you get stuck at this problem I can try debugging it too, but I don't want to do any double work.

Comment (NLBuescher) [inline lib/tag-finder.coffee]: I noticed that the first time around, but I copied the sequence again, and made sure to escape _all_ the slashes (I used a regex tester and it told me to escape even the `/`). That seems to have fixed the problem. Thanks for you help. 

I'll go ahead and push the changes and write a spec for the new cases.

**Comment by NLBuescher** [comment]
So I did a rebase with upstream, and the specs ran fine on my machine (doesn't mean much, I know), but now one of the checks isn't working, but it's only the x64 version of the stable channel check that has trouble with AppVeyor. So it doesn't seem to be any obvious problem with the code I added.

Does anyone have any advice on what I can do to fix this? I can try to revert to rebase if that will help.

**Comment by 50Wliu** [comment]
Do not worry about CI unless it looks like a spec/linting error that you caused.

However, it does look like you messed up the rebase, since there's a lot of commits that are already on master that are showing up on this branch.

**Comment by NLBuescher** [comment]
Alright. I'll try reverting the rebase.

Should I try to rebase at all, or is that not necessary?

**Comment by 50Wliu** [comment]
You don't need to if there's no conflicts.

**Comment by NLBuescher** [comment]
Alright, rebase has been reverted. As a bonus, now all the specs pass, so there's that.

**Comment by ungb** [comment]
was able to verify:

Before:

image


After:
image



pr https://github.com/atom/bracket-matcher/pull/273 fixes the issue with completing closing tag.

**Comment by as-cii** [comment]
Thanks for this pull request, @NLBuescher! Apologies for the delay in answering this. 🙇 

Could you resolve the conflicts with master so that I can proceed with reviewing this?

**Comment by NLBuescher** [comment]
@as-cii I've resolved the conflicts.

**Comment by ruv** [comment]
@NLBuescher, Why don't you rebase self-closing-tag-fix branch on master instead of merging master into this branch?

**Comment by NLBuescher** [comment]
@ruv I just clicked on the button, here on the website, that said "Resolve Conflicts", edited the conflicting lines in the web editor, clicked on commit changes, and it did the thing. I messed up a rebase before, so I'm just glad it works without adding a massive amount of commit messages.

**Comment by NLBuescher** [comment]
@as-cii It's been a while since anything has happened on this thread. Is there anything else I need to do?

**Comment by lee-dohm** [comment]
Thanks for your perseverance @NLBuescher! Great work :bow:

**Comment by ungb** [comment]
I was able to test this after the merge conflict was resolve and see that the issue with self closing tags is fixed.

**Comment by mqklin** [comment]
I installed it through `apm install bracket-matcher@0.87.1` and after one tags matching atom hang up. So I can't use this feature now, made `apm uninstall bracket-matcher`. When this feature will be available?

**Comment by ungb** [comment]
@mqklin, can you provide more info on the file you're making edits to, your os info, and atom version  ?

 i'll take a look.  apm list to show what other community  packages you have installed could be helpful too.

I haven't seen atom hang but would be good to get repro steps to understand what you were doing when you saw this issue

**Comment by ungb** [comment]
@mqklin i was able to repro the issue you saw and created issue https://github.com/atom/bracket-matcher/issues/303.  I've already bumped back the version on Atom master to not include this change, but will be looking to revert this change soon as soon as I figure out the merge conflicts.  

@NLBuescher I'm not sure if you have time to take a look, but looks like this PR actually causes Atom to hang.  If you take a look at issue https://github.com/atom/bracket-matcher/issues/303 for repro steps.

**Comment by NLBuescher** [comment]
@ungb I'll take a look as soon as I can.
